File: Ruby_version/Napakalaki/lib/cultist_player.rb
# encoding: utf-8
# begin cultist_player.rb

require_relative 'player'
require_relative 'cultist'

module NapakalakiGame
  class CultistPlayer < Player
    
    # Nº de jugadores sectarios en la partida
    @@totalCultistPlayers = 0
    
    def initialize(p, c)
      # Llamar constructor de arriba ??? => super pero ....
    
      @@totalCultistPlayers += 1
      @myCultistCard = c
    end
    
    def self.getTotalCultistPlayers
      @@totalCultistPlayers
    end
    
    def self.totalCultistPlayers
      @@totalCultistPlayers
    end
    
    def getCombatLevel
      (super * 1.7 + @myCultistCard.getGainedLevels * @@totalCultistPlayers).round
    end
    
    def getOponentLevel(m)
      m.getCombatLevelAgainstCultistPlayer
    end
    
    def shouldConvert
      false
    end
    
    def giveMeATreasure
      @visibleTreasures.delete(@visibleTreasures.sample)
    end
    
    def canYouGiveMeATreasure
      !getHiddenTreasures.empty?
    end
    
    protected :getCombatLevel
    protected :getOponentLevel
    protected :shouldConvert
    private :giveMeATreasure
    private :canYouGiveMeATreasure
    
  end
end

# end cultist_player.rb

File: Java_version/Napakalaki/src/NapakalakiGame/CultistPlayer.java
// begin CultistPlayer.java
package NapakalakiGame;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.Random;
import java.lang.Math;

public class CultistPlayer extends Player {
    private static int totalCultistPlayers = 0;
    Cultist myCultistCard;
    
    public CultistPlayer(Player p, Cultist c) {
        super(p);
        myCultistCard = c;
        totalCultistPlayers++;
    }
    
    @Override
    public int getCombatLevel() {
        return (int)Math.round(1.7*super.getCombatLevel() + 
                myCultistCard.getGainedLevels()*totalCultistPlayers);
    }
    
    @Override
    protected int getOponentLevel(Monster m) {
        return m.getCombatLevelAgainstCultistPlayer();
    }

    @Override
    protected boolean shouldConvert() {
        return false;
    }
    
    @Override
    protected Treasure giveMeATreasure() {
        Treasure t = visibleTreasures.get(new Random().nextInt(
            visibleTreasures.size()));
        visibleTreasures.remove(t);
        return t;
    }
    
    @Override
    protected boolean canYouGiveMeATreasure(){
        return !visibleTreasures.isEmpty();
    }
    
    static public int getTotalCultistPlayers() {
        return totalCultistPlayers;
    }
    
    @Override
    public String toString() {
        return super.toString() + "\n I´m a cultist.";
    }
}

// end CultistPlayer.java